Back swift_task_enqueueGlobal with public Dispatch API

rjmccall opened the pull request:
We expect to iterate on this quite a bit, both publicly and internally, but this is a fine starting-point.
    
I've renamed `runAsync` to runAsyncAndBlock to underline very clearly what it does and why it's not long for this world.  I've also had to give it a radically different implementation in an effort to make it continue to work given an actor implementation that is no longer just running all work synchronously.
    
The major remaining bit of actor-scheduling work is to make `swift_task_enqueue` actually do something sensible based on the executor it's been given; currently it's expecting a flag that IRGen simply doesn't know to set.  But we may want to express this as a more general actor-to-executor projection step.

rjmccall committed:
[NFC] Add a header library to simplify generating synthetic types

I'm going to use this in Builtins.cpp.

rjmccall committed:
Add a builtin to convert a Task* to a Job*.

rjmccall committed:
Use current public Dispatch API to schedule global work.

We expect to iterate on this quite a bit, both publicly
and internally, but this is a fine starting-point.

I've renamed runAsync to runAsyncAndBlock to underline
very clearly what it does and why it's not long for this
world.  I've also had to give it a radically different
implementation in an effort to make it continue to work
given an actor implementation that is no longer just
running all work synchronously.

The major remaining bit of actor-scheduling work is to
make swift_task_enqueue actually do something sensible
based on the executor it's been given; currently it's
expecting a flag that IRGen simply doesn't know to set.

rjmccall commented:
@swift-ci Please test

DougGregor commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -1357,21 +1359,64 @@ void swift::swift_task_enqueue(Job *job, ExecutorRef executor) {
+  // contrast, if the async feature adds too much overhead, or if
+  // heavy use of it leads to thread explosions and memory exhaustion,
+  // programmers will have no choice but to stop using it.  So if
+  // goals are in conflict, it's best
```
Comment tails off here

rjmccall replied:
Fixed

DougGregor commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -137,11 +137,6 @@ extension Task {
       Builtin.cancelAsyncTask(task)
     }
-
-    @available(*, deprecated, message: "This is a temporary hack")
```
Yay!

ktoso replied:
Wohoo! Can't wait

DougGregor commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -308,6 +306,9 @@ extension Task {
     let (task, _) = Builtin.createAsyncTaskFuture(flags.bits, nil, operation)
 
+    // Enqueue the resulting job.
+    _enqueueJobGlobal(Builtin.convertTaskToJob(task))
```
We'll want to do this for `_runChildTask` as well.

DougGregor commented on CMakeLists.txt:
```diff
@@ -941,7 +941,7 @@ if (LLVM_ENABLE_DOXYGEN)
 endif()
 
-if(SWIFT_BUILD_SYNTAXPARSERLIB OR SWIFT_BUILD_SOURCEKIT)
+if(SWIFT_BUILD_SYNTAXPARSERLIB OR SWIFT_BUILD_SOURCEKIT OR SWIFT_ENABLE_EXPERIMENTAL_CONCURRENCY)
```
I'm guessing you'll need to teach `_Concurrency` to link against the (C) lib dispatch on Linux and Windows.

rjmccall committed:
Add builtins to initialize and destroy a default-actor member.

It would be more abstractly correct if this got DI support so
that we destroy the member if the constructor terminates
abnormally, but we can get to that later.

rjmccall commented:
@swift-ci Please test

DougGregor commented:
Windows failure:
```
/Fostdlib\public\Concurrency\CMakeFiles\swift_Concurrency-windows-x86_64.dir\Actor.cpp.obj /Fdstdlib\public\Concurrency\CMakeFiles\swift_Concurrency-windows-x86_64.dir\ /FS -c S:\jenkins\workspace\swift-PR-windows\swift\stdlib\public\Concurrency\Actor.cpp
S:\jenkins\workspace\swift-PR-windows\swift\stdlib\public\Concurrency\Actor.cpp(1419,43): error: use of undeclared identifier 'qos_class_t'
  auto queue = dispatch_get_global_queue((qos_class_t) priority,
                                          ^
```

DougGregor commented:
Okay, I think the Windows failure is because you need to use `dispatch_qos_class_t` instead of `qos_class_t`. `qos_class_t` seems like the Darwin one.

DougGregor commented on include/swift/AST/Builtins.def:
```diff
@@ -718,6 +723,12 @@ BUILTIN_MISC_OPERATION(TypePtrAuthDiscriminator, "typePtrAuthDiscriminator", "n"
+BUILTIN_MISC_OPERATION(InitializeDefaultActor, "initializeDefaultActor", "", Special)
+
+/// Destroy the default-actor instance in a default actor object.
+BUILTIN_MISC_OPERATION(DestroyDefaultActor, "destroyDefaultActor", "", Special)
```
I think you'll need to add all of these to `visitBuiltinAddress` in `MemAccessUtils.cpp`. There's a non-exhaustive switch in there that's bitten me a few times.

rjmccall replied:
@atrick Andy, can you clarify whether it's necessary/correct to update MemAccessUtils.cpp?  It looks like it's visiting uses of addresses.  The default-actor state is part of the class instance, but it's not otherwise modeled in SIL at all.

atrick replied:
@rjmccall  It looks like you added builtins that take a reference operand. Those will look like any other reference-escaping operation. MemAccessUtils won't try to build an access path for those, but they will look like unknown uses of the object that may overlap with other access to the object. If you want to make sure alias analysis disambiguates those from property access, then you can file a performance radar for that general problem. `AccessPathDefUseTraversal::visitSingleValueUser` will need to be updated, but so will MemBehavior etc.

rjmccall replied:
Okay, it's probably not high-priority, but filed as rdar://72170139.

DougGregor commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -1357,21 +1359,66 @@ void swift::swift_task_enqueue(Job *job, ExecutorRef executor) {
+  // the priorities of work added to this queue using Dispatch's public
+  // API, but as discussed above, that is less important than avoiding
+  // performance problems.
+  dispatch_function_t dispatchFunction = &__swift_run_job;
```
Should we separate out the dispatch-interfacing parts of this in their own source file? That seems like the part most likely to get replaced for different platforms, whereas the rest of the Actor implementation will remain similar.

rjmccall replied:
Sure, that's a good idea.

DougGregor commented on stdlib/public/Concurrency/Actor.cpp:
```diff
@@ -1357,21 +1359,66 @@ void swift::swift_task_enqueue(Job *job, ExecutorRef executor) {
+  JobPriority priority = job->getPriority();
+
+  // TODO: cache this to avoid the extra call
+  auto queue = dispatch_get_global_queue((qos_class_t) priority,
```
As I noted elsewhere, I think this needs to be `dispatch_qos_class_t`.

rjmccall replied:
Yeah, I saw that and have the fix already; I'll update the PR as soon as this round of build comes back.

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall committed:
Implement a cooperative global executor for single-threaded runtimes.

Currently, the only thing in the system that donates a thread
to run it is swift_runAndBlockThread, but we'll probably need
others.  Nothing in the concurrency runtime should block via a
semaphore in this configuration.

As an outrageous hack, work around the layering problems with
using libdispatch from the concurrency library on non-Darwin
systems by making those systems use the cooperative global
executor.  This is only acceptable as a temporary solution
for landing this change and setting things onto the right
long-term design.

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
Okay, there's a remaining test failure on Windows, but a bunch of work is blocking on this, so I marked the test as unsupported and filed rdar://72171028 to fix it post-commit.

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test Windows

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test
